akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -109,21 +110,80 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
+      // TODO: make me configurable
+      private val streamTimeout = 100.milliseconds
+      private val connectionTimeout = streamTimeout.plus(streamTimeout)
 
       def lastStreamId: Int = {
         incomingStreams.lastOption.map(_._1).getOrElse(0)
       }
 
-      def pushGOAWAY(errorCode: ErrorCode = Http2Protocol.ErrorCode.PROTOCOL_ERROR): Unit = {
-        // http://httpwg.org/specs/rfc7540.html#rfc.section.6.8
-        val last = lastStreamId
-        closedAfter = Some(last)
-        bufferedFrameOut.push(GoAwayFrame(last, errorCode))
-        // FIXME: handle the connection closing according to the specification
+      def connectionClose(): Unit = {
+        // FIXME: I'm not sure but this does not look working, any ideas?
+        // if this is not propagated we somehow need a way to kill the tcp connection from the demuxer?
+        debug("Closing the connection!")
+        completeStage()
+      }
+
+      def completeStream(stream: SubStream): Unit = {
+        incomingStreams(stream.streamId).state = StreamState.Closed
+        // WINDOW_UPDATE or RST_STREAM frames can be received in this state for a short period after a DATA or HEADERS
+        // frame containing an END_STREAM flag is sent. Until the remote peer receives and processes RST_STREAM or the
+        // frame bearing the END_STREAM flag, it might send frames of these types. Endpoints MUST ignore WINDOW_UPDATE
+        // or RST_STREAM frames received in this state, though endpoints MAY choose to treat frames that arrive a
+        // significant time after sending END_STREAM as a connection error (Section 5.4.1) of type PROTOCOL_ERROR.
+        materializer.scheduleOnce(streamTimeout, new Runnable {
+          override def run(): Unit = {
+            stream.inlet.foreach(_.cancel())
+            stream.outlet.complete()
+            incomingStreams.remove(stream.streamId)
+          }
+        })
+      }
+
+      def pushGOAWAY(errorCode: ErrorCode): Unit = {
+        def shutdownGraceful(): Unit = {
+          // waits for the termination of all substreams
+          if (incomingStreams.isEmpty) connectionClose()
+          else materializer.scheduleOnce(connectionTimeout, new Runnable {
+            override def run(): Unit = shutdownGraceful()
+          })
+        }
+        if (closing == 0) {
+          closing += 1
+          val last = lastStreamId
+          closedAfter = Some(last)
+          bufferedFrameOut.push(GoAwayFrame(last, errorCode))
+          // A GOAWAY frame might not immediately precede closing of the connection; a receiver of a GOAWAY that has
+          // no more use for the connection SHOULD still send a GOAWAY frame before terminating the connection.
+          materializer.scheduleOnce(connectionTimeout, new Runnable {
+            override def run(): Unit = {
+              println(s"Graceful Shutdown ${incomingStreams.isEmpty}")
+              incomingStreams.foreach { case (id, _) ⇒ pushRST_STREAM(id, ErrorCode.CANCEL) }
```

Comment:
Actually a comment here is useful, it actually helps to terminate long running streams